File: renju/src/main/scala/renju/notation/Pos.scala
package renju.notation

final case class Pos(row: Int, col: Int) {

  def idx: Int = this.row * Renju.BOARD_WIDTH + this.col

  override def toString: String = f"${(col + 97).toChar}${row + 1}"

}

object Pos {

  @inline def idxToRow(idx: Int): Int = idx / Renju.BOARD_WIDTH
  @inline def idxToCol(idx: Int): Int = idx % Renju.BOARD_WIDTH

  @inline def rowColToIdx(row: Int, col: Int): Int = row * Renju.BOARD_WIDTH + col

  def fromCartesian(raw: String): Option[Pos] = for {
    row <- raw.drop(1).toIntOption
    col = raw.head
    pos <- this.fromCartesian(col, row)
  } yield pos

  def fromCartesian(rawCol: String, rawRow: Int): Option[Pos] = this.fromCartesian(rawCol.charAt(0), rawRow)

  def fromCartesian(rawCol: Char, rawRow: Int): Option[Pos] = {
    val col = rawCol.toUpper - 65
    val row = rawRow - 1

    Option.when(col >= 0 && row >= 0 && row < Renju.BOARD_WIDTH && col < Renju.BOARD_WIDTH) { new Pos(row, col) }
  }

  def fromIdx(idx: Int): Pos = new Pos(idxToRow(idx), idxToCol(idx))

}

File: rusty_renju_jni/src/main/java/com/do1phin/rustyrenju/Pos.java
package com.do1phin.rustyrenju;

public class Pos {

    public static int cartesianToIdx(int row, int col) {
        return row * Geometry::BOARD_WIDTH + col;
    }

    public static int idxToRow(int idx) {
        return idx / Geometry::BOARD_WIDTH;
    }

    public static int idxToCol(int idx) {
        return idx % Geometry::BOARD_WIDTH;
    }

    private byte value;

    public Pos(int idx) {
        this.value = (byte) idx;
    }

    public Pos(int row, int col) {
        final int idx = cartesianToIdx(row, col);
        this.value = (byte) idx;
    }

    public byte getValue() {
        return this.value;
    }

    public int getIdx() {
        return this.value & 0xFF;
    }

    public int getRow() {
        return idxToRow(this.getIdx());
    }

    public int getCol() {
        return idxToCol(this.getCol());
    }

}
